fix(dynamic-import-vars): keep dynamic import assert arg intact

jorenbroekema opened the pull request:
<!--
  ⚡️ katchow! We ❤️ Pull Requests!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Pull Request Requirements:
  * Please include tests to illustrate the problem this PR resolves.
  * Please lint your changes by running `npm run lint` before creating a PR.
  * Please update the documentation in `/docs` where necessary

  Please place an x (no spaces - [x]) in all [ ] that apply.
-->

<!-- the plugin(s) this PR is for -->

## Rollup Plugin Name: `dynamic-import-vars`

This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

If yes, then include "BREAKING CHANGES:" in the first commit message body, followed by a description of what is breaking. 

List any relevant issue numbers: https://github.com/rollup/plugins/issues/1503

### Description

When transforming dynamic imports with variables in them, the second argument (import assertions) is stripped away and this information is lost. This Pull Request addresses that problem (see issue) by grabbing the argument Node and stringifying it and injecting it into the Magic String, so that this argument is kept as is.

This is needed for other plugins to correctly deal with imports that have import assertions, most notably https://github.com/jleeson/rollup-plugin-import-css which for CSS imports returns either a CSS string or CSSStyleSheet depending on whether or not an import assertion for CSS is present. If such an assertion is present, but dynamic-import-vars strips it away, the CSS plugin will bundle the CSS import as a CSS string rather than CSSStyleSheet, which would be different from local development, before bundling, since browsers natively do support import assertions now and return a CSSStyleSheet.

I might need some help with the testing, as it currently doesn't verify that the assertion is kept intact, I'm not sure how to test that, it seems in the final output[0].code the assertions are gone again, but when I test my changes in my local project, it does fix my issue.. and when I console.log the transform hook in this plugin with my changes, it does seem like it returns the correct code.

shellscape commented on packages/dynamic-import-vars/package-lock.json:
```diff
@@ -0,0 +1,354 @@
+{
```
this file shouldn't exist

jorenbroekema replied:
removed, I probably accidentally installed with npm instead of pnpm or something

lukastaegert reviewed (commented):
Looks mostly fine to me, but the comment about adding the acorn plugin makes me curious as this plugin is already added by Rollup.

shellscape commented:
@jorenbroekema please give the replies here a look 🙏  if we reach 60 days without a reply, we'll consider the PR to be abandoned.

jorenbroekema commented on packages/dynamic-import-vars/README.md:
````diff
@@ -145,6 +145,29 @@ function importLocale(locale) {
 }
 ```
 
+## Import Assertions
+
+This plugin will keep your import assertions inside dynamic import statements intact, provided that you add the `acorn-import-assertions` syntax plugin to your Rollup config:
+
+```js
+import { importAssertions } from 'acorn-import-assertions';
+import dynamicImportVars from '@rollup/plugin-dynamic-import-vars';
+
+export default {
+  plugins: [dynamicImportVars()],
+  acornInjectPlugins: [importAssertions]
+};
+```
+
+Will ensure the following import assertions remain intact:
+
+```js
+// Refer to rollup-plugin-import-css https://github.com/jleeson/rollup-plugin-import-css
+function importLocale(sheet) {
+  return import(`./styles/${sheet}.css`, { assert: { type: 'css' } });
+}
+```
+
````
You're right, it works without needing to add it manually. I guess I based that off of some other plugin that was based on an older version of Rollup where this was still required.

jorenbroekema commented:
Thanks for the reminder @shellscape , I addressed your comments, I might still need a hand on the test, the snapshot output seems to show the assert arg not appearing in the output, I'm not sure that's intended and I'm missing something, or my test is incorrect?

shellscape reviewed (approved):
I think that test is sufficient for now. Thanks for adding it. @lukastaegert please have another look.

lukastaegert reviewed (approved):
Looks good to me 👍

shellscape commented:
Sorry about the delay getting this merged. Thanks again!
